File: 231121/네 방향 탈출 가능 여부 판별하기/determine-escapableness-with-4-ways.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int n, m;
int arr[100][100];
bool isPass;
queue<pair<int,int>> q;

int dx[4] = {1,-1,0,0};
int dy[4] = {0,0,1,-1};

bool inRange(int x, int y) {
    return 0 <= x && x < n && 0 <= y && y < m;
}

bool CanGo(int x, int y) {
    if(!inRange(x,y)) return false;
    if(arr[x][y] == 0) return false;
    return true;
}

void Push(int x, int y) {
    arr[x][y] = 0;
    q.push(make_pair(x,y));
}

bool Avaliable(int x, int y) {
    return x == n-1 && y == m-1;
}

bool BFS() {
    while(!q.empty()) {
        int x = q.front().first;
        int y = q.front().second;
        q.pop();

        if(Avaliable(x,y)) {
            return true;
        }

        for(int dir = 0; dir < 4; dir++) {
            int nx = x + dx[dir];
            int ny = y + dy[dir];
            if (CanGo(nx,ny))
                Push(nx,ny);
        }
    }
    return false;
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n >> m;
    for(int i = 0; i<n; i++)
        for(int j = 0; j<n; j++)
            cin >> arr[i][j];

    Push(0,0);
    if (BFS()) cout << 1;
    else cout << 0;
    return 0;
}

File: 250220/네 방향 탈출 가능 여부 판별하기/determine-escapableness-with-4-ways.js
class Queue {
    constructor() {
        this.arr = []
        this.head = -1
        this.tail = -1
    }

    empty() {
        return this.head === this.tail
    }

    push(e) {
        this.arr.push(e)
        this.tail++
    }

    pop() {
        if(this.empty()) {
            throw new Error("Queue size is 0")
        }
        return this.arr[++this.head]
    }

    front() {
        if(this.empty())
            throw new Error("Queue size is 0")
        return this.arr[this.head + 1]
    }

    size() {
        return this.tail - this.head
    }
}

const fs = require('fs')
const input = fs.readFileSync(0).toString().trim().split('\n')

const [n,m] = input[0].split(' ').map(Number)
const arr = input.slice(1).map(v => v.split(' ').map(Number))

const visited = Array(n).fill().map(v => Array(m).fill(false))
const q = new Queue()

const InRange = (x, y) => 0 <= x && x < n && 0 <= y && y < m

const CanGo = (x, y) => {
    if(!InRange(x,y)) return false;
    if(visited[x][y]) return false;
    if(!arr[x][y]) return false;
    return true;
}

const BFS = () => {
    const dx = [0,0,1,-1]
    const dy = [1,-1,0,0]

    while(!q.empty()) {
        const [x,y] = q.pop()

        for(let dir = 0; dir < 4; dir++) {
            const nx = x + dx[dir]
            const ny = y + dy[dir]

            if(CanGo(nx,ny)) {
                q.push([nx,ny])
                visited[nx][ny] = true
            }
        }
    }
}

q.push([0,0])
visited[0][0] = true;
BFS()

if(visited[n-1][m-1]) console.log(1)
else console.log(0)